### PR TITLE
Works on adding a group page

### PR DESCRIPTION
**Pull request by vlop11**
The group page will have a list of all available groups that a user can join. Hopefully, this will also work with the Datastore and will allow the user the option to choose and add a group to their account.

### CODE REVIEW

File: src/main/webapp/groupsPage.html [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        <div>
+            <h1 class="heading mdc-typography--headline3">Movements</h1>
+        </div>
+
+        <div class="card-gallery">
+            <div class="mdc-card">
+                <div class="mdc-card__media mdc-card__media--16-9" style="background-image: url(https://images.unsplash.com/photo-1591622414912-34f2a8f8172e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80);"></div>
+                <div class="title-container">
+                    <h2 class="text mdc-typography mdc-typography--headline6">Black Lives Matter</h2>
+                    <!-- <h3 class="demo-card__subtitle mdc-typography mdc-typography--subtitle2">by Kurt Wagner</h3> -->
```
Comment (ahlschlager) [inline src/main/webapp/groupsPage.html]: Are the comments (along with lines 43 and 44) necessary?

File: src/main/webapp/groupsPage.css [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    width: 750px;
+}
+
+.mdc-card {
+    margin: auto;
+    width: 500px;
+}
+
+.text {
+    padding: 0px 15px;
```
Comment (dylancastillo) [inline src/main/webapp/groupsPage.css:16]: nit: [no px](https://google.github.io/styleguide/htmlcssguide.html#0_and_Units) after 0

Comment (dylancastillo) [inline src/main/webapp/groupsPage.css]: nit: [no px](https://google.github.io/styleguide/htmlcssguide.html#0_and_Units) after 0

File: src/main/webapp/groupsPage.html [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                </div>
+
+                <div class="desc-container mdc-typography mdc-typography--body2">
+                    <p class="text">
```
Comment (courtneywong) [inline src/main/webapp/groupsPage.html]: Try to make this selector more specific. `text` is very generic, and you will likely add more text to this page that won't want the same styling.

**Review by courtneywong — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        SoyFileSet sfs = SoyFileSet
+            .builder()
+            .add(new File("../../src/main/java/templates/pages.soy"))
+            .build();
+        SoyTofu tofu = sfs.compileToTofu();
+
+        String out = tofu.newRenderer("templates.pages.groupsPage").setData(data).render();
+        response.getWriter().println(out);
+    }
+}
```
Comment (ahlschlager) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java]: Newline?

File: src/main/java/templates/pages.soy [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                                <i class="material-icons mdc-icon-button__icon">favorite_border</i>
+                            </button>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </body>
+
+    </html>
+{/template}
```
Comment (ahlschlager) [inline src/main/java/templates/pages.soy]: newline

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        data.put("groupDescription", "Group description would go here...");
+
+        SoyFileSet sfs = SoyFileSet
+            .builder()
+            .add(new File("../../src/main/java/templates/pages.soy"))
+            .build();
+        SoyTofu tofu = sfs.compileToTofu();
+
+        String out = tofu.newRenderer("templates.pages.groupsPage").setData(data).render();
+        response.getWriter().println(out);
```
Comment (ahlschlager) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java:44]: What exactly is this line doing? (line 35)

Comment (vlop11) [reply src/main/java/com/google/sps/servlets/GroupsServlet.java:44]: honestly, that was just in the tutorial but from what I understand I pass in my data (that's the data map) and it'll render the page

File: src/main/webapp/groupsPage.html [hunk 7]
```diff
@@ -0,0 +1,47 @@
+<!DOCTYPE html>
```
Comment (courtneywong) [inline src/main/webapp/groupsPage.html]: Why do we need this page when we have the Soy template? See if my comments on https://github.com/googleinterns/step123-2020/pull/10 are relevant here.

**Review by courtneywong — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 8]
```diff
@@ -14,24 +14,28 @@
...
 public class GroupsServlet extends HttpServlet {
     
+    /**
+     * Renders a HTML template with cards representing the various
+     * groups that have already been created.
+     */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
         Map<String, String> data = new HashMap<>();
```
Comment (ahlschlager) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java]: Is this hashmap only temporary? If not are you thinking of making it immutable?

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        ImmutableList.Builder<ImmutableMap<String, String>> builder = new ImmutableList.Builder<>();
+
+        // Hard coded data for now but will use data from datastore
+        // as groups are added to datastore
+        ImmutableMap<String, String> groupMap = ImmutableMap.of("groupName", "Black Lives Matter",
+            "groupImage", "https://images.unsplash.com/photo-1591622414912-34f2a8f8172e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
+            "groupDescription", "Group description would go here...");
+        builder.add(groupMap);
+        ImmutableList<ImmutableMap<String, String>> groups = builder.build();
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java]: You can shorten this to `groups = ImmutableList.of(groupMap)` and also remove line 24.

File: src/main/java/templates/groups.soy [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+                       <div class="mdc-card__action-icons">
+                           <button class="mdc-icon-button mdc-card__action mdc-card__action--icon--unbounded" aria-pressed="false" aria-label="Add to favorites" title="Add to favorites">
+                               <i class="material-icons mdc-icon-button__icon mdc-icon-button__icon--on">favorite</i>
+                               <i class="material-icons mdc-icon-button__icon">favorite_border</i>
+                           </button>
+                       </div>
+                   </div>
+               </div>
+               {/for}
```
Comment (courtneywong) [inline src/main/java/templates/groups.soy]: Can you fix the indentation here so that everything is nested correctly?

File: src/main/java/templates/groups.soy [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+           </div>
+
```
Comment (courtneywong) [inline src/main/java/templates/groups.soy:32]: nit: remove newline

File: src/main/java/templates/groups.soy [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        </body>
+
```
Comment (courtneywong) [inline src/main/java/templates/groups.soy]: nit: remove newline

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+     * groups that have already been created.
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        ImmutableList.Builder<ImmutableMap<String, String>> builder = new ImmutableList.Builder<>();
+
+        // Hard coded data for now but will use data from datastore
+        // as groups are added to datastore
+        ImmutableMap<String, String> groupMap = ImmutableMap.of("groupName", "Black Lives Matter",
+            "groupImage", "https://images.unsplash.com/photo-1591622414912-34f2a8f8172e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java:29]: For images that are being used, you should make sure that they can be used for commercial use. You should also give credit somewhere to images that you're using.

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+            "groupDescription", "Group description would go here...");
+        builder.add(groupMap);
+        ImmutableList<ImmutableMap<String, String>> groups = builder.build();
+        
+        // Each group has its own map which points to its info and all maps are passed into the template as a list
+        // This will make it easier when groups are queried from Datastore
+        ImmutableMap<String, ImmutableList<ImmutableMap<String, String>>> data = ImmutableMap.of("groups", groups);
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java:35]: Can you make a Group object instead of a Map?

**Review by courtneywong — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroupsServlet.java (hunk 13 above)
Comment (vlop11) [reply src/main/java/com/google/sps/servlets/GroupsServlet.java:29]: I've used Unsplash so far which allows commercial use, and you don't even have to give credit (though I plan to anyway--probably when we do styling I'll add a little text box)

File: src/main/java/com/google/sps/servlets/GroupsServlet.java (hunk 14 above)
Comment (vlop11) [reply src/main/java/com/google/sps/servlets/GroupsServlet.java:35]: Can java classes be passed into a template? I looked into it, but there wasn't a whole lot of information. How do you think I should do it?

File: src/main/java/templates/groups.soy [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                            </div>
+
+                            <div class="mdc-card__action-icons">
+                                <button class="mdc-icon-button mdc-card__action mdc-card__action--icon--unbounded" aria-pressed="false" aria-label="Add to favorites" title="Add to favorites">
+                                    <i class="material-icons mdc-icon-button__icon mdc-icon-button__icon--on">favorite</i>
+                                    <i class="material-icons mdc-icon-button__icon">favorite_border</i>
+                                </button>
+                            </div>
+                        </div>
+                    </div>
```
Comment (courtneywong) [inline src/main/java/templates/groups.soy]: Extract this out to a separate Group template. I'd recommend having the group template take the params `image` `name` `description` to make it clearer what information is being passed around.

File: src/main/java/com/google/sps/servlets/GroupsServlet.java (hunk 14 above)
Comment (courtneywong) [reply src/main/java/com/google/sps/servlets/GroupsServlet.java:35]: Ah, we typically pass protos around, but it doesn't seem like that's supported in the open source version of Soy.

**Review by courtneywong — CHANGES_REQUESTED** [review]

**Review by ahlschlager — APPROVED** [review]

**Review by courtneywong — APPROVED** [review]

**Review by sporkert — APPROVED** [review]
